sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -536,6 +536,11 @@ object TypeCoercion {
           case None => c
         }
 
+      case ArrayJoin(arr, d, nr) if !ArrayType(StringType).acceptsType(arr.dataType) &&
+        ArrayType.acceptsType(arr.dataType) =>
+        val containsNull = arr.dataType.asInstanceOf[ArrayType].containsNull
+        ArrayJoin(Cast(arr, ArrayType(StringType, containsNull)), d, nr)
```

Comment:
not every type can be casted to `StringType`. What about using `ImplicitTypeCasts.implicitCast` in order to check if we can cast it?

Reply:
Hi @mgaido91,
to be honest, I've considered this option before submitting this PR. But I'm glad that you mentioned this approach. At least, we can discuss pros and cons of different solutions. Usage of ```ImplicitTypeCasts.implicitCast``` would enable conversion only from primitive types. I think it would be nice to support non-primitive types as well. WDYT?

Re: Casting to ```StringType```: According to ```Cast.canCast``` method should be possible to cast any type to ```StringType```:
**line 42:** ```   case (_, StringType) => true```
Or am I missing something? I hope test cases in *.../typeCoercion/native/arrayJoin.sql* cover to ```StringType``` conversions from all Spark types.

Reply:
I am not sure. I think it is arguable which is the right result of `SELECT array_join(array(array('a', 'b'), array('c', 'd')), ';')` for instance. With this PR, the result is `[a, b];[c, d]` but shouldn't it be `[a;b];[c;d]`? Moreover, Presto, which is the reference here, doesn't support nested arrays for instance:
```
presto> select array_join(array[array[1, 2, 3], array[3, 4, 5]], ';');
Query 20180625_090549_00003_bsbcg failed: Input type array(integer) not supported
```
So, I'd avoid that.

Reply:
Ok, no problem. Let's support just arrays of primitive types for now. Thanks!